kaldb/src/test/java/com/slack/kaldb/metadata/core/KaldbMetadataStoreTest.java

```diff
@@ -394,4 +394,56 @@ public TestMetadataStore() {
       throw new RuntimeException(e);
     }
   }
+
+  @Test
+  public void testSlowCacheInitialization() {
+    class FastMetadataStore extends KaldbMetadataStore<TestMetadata> {
+      public FastMetadataStore() {
+        super(
+            curatorFramework,
+            CreateMode.PERSISTENT,
+            true,
+            new JacksonModelSerializer<>(TestMetadata.class),
+            STORE_FOLDER);
+      }
+    }
+
+    class SlowSerializer implements ModelSerializer<TestMetadata> {
+      final JacksonModelSerializer<TestMetadata> serializer =
+          new JacksonModelSerializer<>(TestMetadata.class);
+
+      @Override
+      public byte[] serialize(TestMetadata model) {
+        return serializer.serialize(model);
+      }
+
+      @Override
+      public TestMetadata deserialize(byte[] bytes) {
+        try {
+          Thread.sleep(200);
+        } catch (InterruptedException e) {
+          throw new RuntimeException(e);
+        }
+        return serializer.deserialize(bytes);
+      }
+    }
+
+    class SlowMetadataStore extends KaldbMetadataStore<TestMetadata> {
+      public SlowMetadataStore() {
+        super(curatorFramework, CreateMode.PERSISTENT, true, new SlowSerializer(), STORE_FOLDER);
+      }
+    }
+
+    int testMetadataInitCount = 10;
+    try (KaldbMetadataStore<TestMetadata> init = new FastMetadataStore()) {
+      for (int i = 0; i < testMetadataInitCount; i++) {
+        init.createSync(new TestMetadata("name" + i, "value" + i));
+      }
+    }
+
+    try (KaldbMetadataStore<TestMetadata> init = new SlowMetadataStore()) {
+      List<TestMetadata> metadata = init.listSync();
+      assertThat(metadata.size()).isEqualTo(testMetadataInitCount);
```

Comment:
cool! this assert fails when I commented out `awaitCacheInitialized` from the listAsync method :)